### PR TITLE
Add support for an ancestor quick fix

### PR DESCRIPTION
**Pull request by tbouron**
This will check if an entity has implicit inherited config keys. If this is the case, then warnings will be created for these config keys. Additionally, a quick fix will be available to automatically set the implicit inherited config key `x` to `$brooklyn:parent().config("x")`

Here it is in action:
quick-fix-ancestor

### CODE REVIEW

File: ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html [hunk 1]
```diff
@@ -19,129 +19,258 @@
...
+                            <div ng-repeat="fix in itemV.quickFixes" class="error-line-sub-line">
+                                <div class="error-line-marker">
+                                    <i class="fa fa-fw fa-magic"></i>
+                                </div>
+                                <div class="error-line-text">
+                                    {{ fix.text }}
+                                    <a class="hand btn btn-xs btn-primary" style="float: right;" ng-click="vm.applyQuickFix(fix)"
+                                       ng-attr-title="{{ fix.tooltip }}"
+                                    >Apply
+                                        (<ng-pluralize count="vm.size(fix.issues)"
```
Comment (algairim) [inline ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:196]: Typo: redundant parentheses at the beginning.

Comment (algairim) [reply ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:196]: Or make parentheses at the end visible. It is not displayed.

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js [hunk 2]
```diff
@@ -55,6 +55,31 @@ export function computeQuickFixes(blueprintService, allIssues) {
...
+                v = allIssues.warnings.byMessage[key] = {
+                    group: issue.group,
+                    ref: issue.ref,
+                    message: issue.message,
+                    issues: [],
+                    quickFixes: {},
+                };
+            }
+
+            let issueO = {
```
Comment (algairim) [inline ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:74]: Is it `issueObject`?

File: ui-modules/blueprint-composer/app/components/providers/blueprint-service.provider.js [hunk 3]
```diff
@@ -554,6 +564,24 @@ function BlueprintService($log, $q, $sce, paletteApi, iconGenerator, dslService,
...
+        return $q((resolve) => {
+            entity.miscData.get('config')
+                .filter(definition => !entity.config.has(definition.name))
+                .forEach(definition => {
+                    if (entity.hasInheritedConfig(definition.name)) {
+                        entity.addIssue(Issue.builder()
+                            .group('config')
+                            .ref(definition.name)
+                            .level(ISSUE_LEVEL.WARN)
+                            .message(`Implicitly defined from one of it's ancestor`)
```
Comment (algairim) [inline ui-modules/blueprint-composer/app/components/providers/blueprint-service.provider.js:577]: typo: _...one of its ancestors_.

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js [hunk 4]
```diff
@@ -55,6 +55,31 @@ export function computeQuickFixes(blueprintService, allIssues) {
...
+                issue,
+                //quickFixes: {},
+            }
+            v.issues.push(issueO);
+
+            computeQuickFixesForIssue(issue, issue.entity, blueprintService, v.quickFixes)
+        });
+    });
```
Comment (algairim) [inline ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:82]: Duplicate code, can be moved into a function like:
```
export function computeQuickFixes(blueprintService, allIssues) {
    if (!allIssues) allIssues = blueprintService.getAllIssues();
    if (!allIssues) allIssues = {};
    if (!allIssues.errors) allIssues.errors = {};

    computeQuickFixesForIssues(allIssues.errors, blueprintService);
    computeQuickFixesForIssues(allIssues.warnings, blueprintService);
    
    return allIssues;
}

function computeQuickFixesForIssues(issuesCollection, blueprintService) {
    issuesCollection.byMessage = {};
    Object.values(issuesCollection.byEntity).forEach(list => {

    // etc.
```

File: ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html (hunk 1 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:196]: @algairim It is at the end of the `ng-pluralize` directive, on line 197

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js (hunk 2 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:74]: This is a copy-paste from existing code, see more detailled explanation below @algairim

File: ui-modules/blueprint-composer/app/components/providers/blueprint-service.provider.js (hunk 3 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/components/providers/blueprint-service.provider.js:577]: Good catch, I'll update that @algairim

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js (hunk 4 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:82]: @algairim Yeah about that. This was the quickest and dirtiest way to make this work (the previous version supported only warning) 

However, the way the quick fixes are implemented doesn't really follow the AngularJS best practices or "way". I plan du refactor that which would make this bit of code go away. That's why I didn't go for a DRY implementation just now.

File: ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html (hunk 1 above)
Comment (algairim) [reply ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:196]: Screenshot 2021-07-09 at 09 13 46
parentheses  at 197 is not displayed.

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js [hunk 5]
```diff
@@ -353,7 +398,7 @@ export function getQuickFixProposer(type) {
 export function getQuickFixHintsForIssue(issue, entity) {
     if (issue.group === 'config') {
         let hints = (entity.miscData.get('ui-composer-hints') || {})['config-quick-fixes'] || [];
-        hints = hints.filter(h => h.key === issue.ref);
+        hints = hints.filter(h => new RegExp(h.key).test(issue.ref));
```
Comment (algairim) [inline ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:401]: Puzzling myself with this RegExp. It supposed to take regex expression as a parameter, however, `h.key` is a plain string if I understand that correctly. Does it work? What do we filter here?

File: ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html (hunk 1 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:205]: Urgh, forgot that AngularJS is picky with self-closing tag. It's fixed now

Comment (algairim) [reply ui-modules/blueprint-composer/app/views/main/graphical/graphical.state.html:205]: Now it works, thanks @tbouron

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js (hunk 4 above)
Comment (algairim) [reply ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:82]: Ok, a little effort to move this code into function though :)

File: ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js (hunk 5 above)
Comment (tbouron) [reply ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:401]: @algairim `RegExp` can take a regex or a string as regex, this is perfectly valid. The idea is to filter for the specific config key but in the case of implicit use of inherited config key, we want to do the check for ALL config key (see: https://github.com/apache/brooklyn-ui/pull/241/files#diff-b9b9776bc3d54d479c21091094606ac270b807327ee31a93317a3598964e2749R39)

This is merly to be backward compatible with what was there previously. The quick fix API is super obscure and I plan to surface this kind of thing much more when refactoring it

Comment (algairim) [reply ui-modules/blueprint-composer/app/components/quick-fix/quick-fix.js:401]: I see, thanks for clarifying.

**Comment by algairim** [comment]
Reminder to myself: @algairim to fix overlapping of multiple labels on relationship arcs:
Screenshot 2021-07-09 at 09 29 10

**Comment by algairim** [comment]
Looks good @tbouron

**Review by algairim — APPROVED** [review]

**Review by algairim — CHANGES_REQUESTED** [review]
Wait, spec-editor unit-tests are failing. Seems to be related to these changes.

**Review by algairim — APPROVED** [review]
All tests passed, good to go.

**Pull request merged**